infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/LogReplicationAckReader.java

```diff
@@ -54,7 +54,7 @@ public class LogReplicationAckReader {
     private long lastAckedTimestamp = Address.NON_ADDRESS;
 
     // Sync Type for which last Ack was received. Default to LOG_ENTRY as this is the initial FSM state
-    private SyncType lastSyncType = SyncType.LOG_ENTRY;
+    private SyncType lastSyncType = null;
```

Comment:
Update the comment as well?

Reply:
Now that lastSyncType is null. In the event that replication was ongoing and the node is restarted. When going into 'onEnter' of initializedState, this will be called:

            fsm.getAckReader().markSyncStatus(LogReplication.SyncStatus.STOPPED);

But, lastSyncType at this point is null, so you'll hit a NPE in LogReplicationMetadataManager L.653

I'm wondering if it' better to have a third syncType value which is UNDEFINED or UNAVAILABLE or NOT_SYNCING

Reply:
Also a test can be added for this case. It would be nice that you write the test before the change so you validate it.

Reply:
We had some discussion about introducing a third syncType but thought it wouldn't be worth making a breaking change.

Reply:
Yes the NPE issue upon node restart does exist. I have 2 doubts here:

1. A bit further for the initial value of `lastSyncType`, why we are giving default value for this in-memory field? shouldn't it be fetched from status table on node start?
2. For protobuf enum, I also think there is some deficiency in our schema, not only causing confusion in this fix. We are widely using 0-indexed enum definition there, but 0 is the default value as per proto 3 semantics. Default value could be given silently and is not printed out in the log. That means a more reasonable enum should be like
```
  enum SyncType {
    UNDEFINED = 0;
    SNAPSHOT = 1;
    LOG_ENTRY = 2;
  }
 ```
 but yes this is a breaking change and will break backward compatibility...
 
 For a quick fix of this PR, would it be okay if we just wrap up the MetadataManager's clause with `lastSyncType != null`?

Reply:
1. I don't remember if on V2 we initialize the value on session creation (I think yes, but better confirm) but not on V1. BTW, that reminds me that this change needs to be on master. 

2. Correct, based on these default values I brought up the SnapshotSyncInfo part. And yes you're right unfortunately, we can't do this change (default behavior) we could only do something to explicitly set it. 

And if it's null what do you do?

Reply:
I believe the NPE will not be hit in the case of a restart because onEntry() for Initialized state will not be executed then.

But I agree with the current limitation of not having an init enum with index 0.

Reply:
@xcchang I do agree with the approach, but just to be clear the summary is not accurate as the issue is that though in this fix we're setting the ReplicationStatusInfo to the correct value (outer), for the case of `snapshot_sync` we don't have enough info to set the inner part SnapshotSyncInfo, so it seems incorrect to set it to 'ONGOING' when that field is not present. Based on that it seems better that : (a) Periodic Task thread only update the status when it is already present in the table (avoiding that first initialization where info is not available) that way it will always pick what was correctly populated by the InSnapshotSyncState or InLogEntrySyncState or (b) what you described, which is, when the value is not present put a 'default' state which reflects that it's not initialized basically.. and then the proper update will come from the state where replication happens. 

@pankti-m why do you say onEntry won't be executed? if this is the first state when the FSM is started.

Reply:
@annym @pankti-m and I has discussed this today, since the onEntry method is called by the "to" state in the transition method of LogReplicationFSM, wouldn't the lastSyncStatus only be null in case of a transition from initialized to initialized?

Reply:
Oh ok, I see your point, true. Though the FSM starts off in this state, because it's not the result of a transition, the 'onEntry' won't be executed. So, I agree null case won't occur. Still, we'd have the issue of incomplete status being set in between initialized state and actual replicating state, right? Because that info is not available for the poller thread.